conans/model/conf.py

```diff
@@ -47,117 +47,404 @@ def _is_profile_module(module_name):
+            new_value = self._value[:]  # do a copy
+            new_value[index:index + 1] = other._value  # replace the placeholder
+            self._value = new_value
+
+
+class Conf:
 
     def __init__(self):
-        self._values = {}  # property: value
+        # It being ordered allows for Windows case-insensitive composition
+        self._values = OrderedDict()  # {var_name: [] of values, including separators}
+
+    def __bool__(self):
+        return bool(self._values)
+
+    # TODO: Python2, remove in 2.0
+    __nonzero__ = __bool__
+
+    def __repr__(self):
+        return "Conf: " + repr(self._values)
+
+    def __eq__(self, other):
+        """
+        :type other: Conf
+        """
+        return other._values == self._values
+
+    # TODO: Python2, remove in 2.0
+    def __ne__(self, other):
+        return not self.__eq__(other)
 
     def __getitem__(self, name):
-        return self._values.get(name)
+        """
+        DEPRECATED: it's going to disappear in Conan 2.0. Use self.get() instead.
+        """
+        # FIXME: Keeping backward compatibility
+        return self.get(name)
 
     def __setitem__(self, name, value):
-        if name != name.lower():
-            raise ConanException("Conf '{}' must be lowercase".format(name))
-        self._values[name] = value
+        """
+        DEPRECATED: it's going to disappear in Conan 2.0.
+        """
+        # FIXME: Keeping backward compatibility
+        self.define(name, value)  # it's like a new definition
 
     def __delitem__(self, name):
+        """
+        DEPRECATED: it's going to disappear in Conan 2.0.
+        """
+        # FIXME: Keeping backward compatibility
         del self._values[name]
 
-    def __repr__(self):
-        return "Conf: " + repr(self._values)
-
     def items(self):
-        return self._values.items()
-
-    def filter_user_modules(self):
-        result = Conf()
+        # FIXME: Keeping backward compatibility
         for k, v in self._values.items():
-            if _is_profile_module(k):
-                result._values[k] = v
-        return result
+            yield k, v.value
 
-    def update(self, other):
+    @property
+    def sha(self):
+        # FIXME: Keeping backward compatibility
+        return self.dumps()
+
+    def get(self, conf_name, default=None, cast=None):
         """
-        :param other: has more priority than current one
-        :type other: Conf
+        Get all the values belonging to the passed conf name. You can convert the final conf value
+        thanks to "cast" param.
+
+        :param conf_name: conf name
+        :param default: default value in case of conf does not have the conf_name key
+        :param cast: any callable to apply any kind of transformation to final conf value
+        """
+        conf_value = self._values.get(conf_name)
+        if conf_value:
+            v = conf_value.value
+            if cast is not None:
+                try:
+                    v = cast(v)
```

Comment:
@memsharded @czoido
In my opinion, this `cast` or `type` (the name can change) is a bit useless. For instance:

```python
# we have this example
conf.get("xxxxxx") == "True"
assert conf.get("xxxxxx", cast=bool) is True  # that's OK

# Now, let's imagine the other case
conf.get("xxxxxx") == "False"
assert conf.get("xxxxxx", cast=bool) is True  # that's OK but weird, isn't it?
```
So, from my point of view, what I expect would be:
```python
# we have this example
conf.get("xxxxxx") == "True"
assert conf.get("xxxxxx", cast=eval) is True  # that's OK

# Now, let's imagine the other case
conf.get("xxxxxx") == "False"
assert conf.get("xxxxxx", cast=eval) is False  # that's OK
```
But it looks a bit odd... What do you think about it? Should this `get-cast` be smarter?

Reply:
Yes, I don't think a cast here is that useful. If we are doing ``parsed_value = eval(value)`` while parsing user input, and that converts directly to Python primitive types, then probably what is better is to remove the ``str()`` conversion done in ``_ConfStrValue`` and move the primitive type (could be directly a boolean) from the user input totally transparently to the user ``conf.get()`` configuration.

Reply:
removed 👍